### DISCUSSION TITLE
Change test case name in XML report

### DISCUSSION THREAD

**Discussion by geokats7** [pytest-dev/pytest#10945] [Q&A]
Is there any way to modify the name of each test which ends up in the XML report (junit) inside the `testcase` tag.
I have managed to use `pytest_collection_modifyitems` hook and change the `_nodeid` but it doesn't seem the best approach.

Is there any other way to access the test case name right before it ends up in the XML report? Through another hook or pytest object maybe...

Following is my example which works but I'd like to avoid it.
```python
def pytest_collection_modifyitems(items):
    for item in items:
        path = item.nodeid.split("::")[0]
        initial_name = item.nodeid.split("::")[1].replace("test_", "")
        name = initial_name.capitalize().replace("_", " ")
        item._nodeid = f"{path}::{name}"
```

**Comment by RonnyPfannschmidt** [comment]
Currently there is not

Note that the approach, you choose can break other tools

Is there any reason why you need to drop test from the name?

**Reply by geokats7** [reply]
That's exactly why I want to avoid it! :D 

My case is this:
We use TestRail's command line tool in order to send the test results AND upload the test cases in TestRail.
This tool uses the XML report produced by pytest at the end of the test session. The names of the test cases which end up in the TestRail are taken straight from the XML report and specifically from the following field,
![image](https://user-images.githubusercontent.com/26500581/234297163-9c034261-1f6c-4582-baa0-6f35e3e75009.png)
and ends up in the TestRail like this,
![image](https://user-images.githubusercontent.com/26500581/234297636-359c69cf-9855-467d-9e61-d9369dee15da.png)

As you guessed, this is not very readable and it would be nice to be able to manipulate the test case name and enter something nicer
(we use the pytest-bdd plugin by the way).

**Reply by RonnyPfannschmidt** [reply]
im not familiar with bdd and testrails

for sanity i would recommend creating a python script that manipulates the xml instead of the ids in pytest

**Reply by geokats7** [reply]
That is an option, yes but it'll be my last resort.

**Reply by RonnyPfannschmidt** [reply]
There currently is no safe /sound way to alter the junit test names, so postprocessing is the best way to change the junit files im currently aware of

**Comment by geokats7** [comment]
I think I found a workaround. I'll leave it here in case someone else has the same issue.
There is pytest fixture called `record_xml_attribute` which lets you modify the _attribute_ of the testcase tag itself thus giving you access to overwrite the test case name.
![image](https://user-images.githubusercontent.com/26500581/234321527-1b865aa1-a829-430a-b961-434d21487b36.png)

You can read more [here](https://docs.pytest.org/en/7.1.x/how-to/output.html#record-xml-attribute).

**Warning**
This feature is experimental, so proceed with caution.

**Reply by RonnyPfannschmidt** [reply]
Please note that you found a bug, this will have to get fixed

**Reply by RonnyPfannschmidt** [reply]
Ok, based on the docs it seems fair to consider the workaround correct as all overrides are allowed
